trinity/tx_pool/validators.py

```diff
@@ -0,0 +1,59 @@
+from typing import (
+    Type
+)
+
+from evm.chains.base import (
+    BaseChain
+)
+from evm.exceptions import (
+    ValidationError
+)
+from evm.rlp.transactions import (
+    BaseTransaction,
+    BaseTransactionFields
+)
+from evm.vm.base import (
+    BaseVM
+)
+
+
+class DefaultTransactionValidator():
+    """
+    The :class:`~trinity.tx_pool.validators.DefaultTransactionValidator` class is responsible to
+    decide wether transactions should be relayed to peers or not. This implementation validates
+    transactions against an initially provided transaction class but will switch to a different
+    one as soon as the tip of the chain uses a more up to date transaction class than the one
+    that was initially provided.
+    """
+
+    def __init__(self, chain: BaseChain, initial_tx_class: Type[BaseTransaction]) -> None:
+        self.chain = chain
```

Comment:
`chain` only appears to be used to `get_vm_configuration`.  What do you think about changing this class to be initialized directly with the` vm_configuration`.

Reply:
We need the chain to get the current VM (or blocknumber theoretically) to validate the tx against the fork at the time of tx relaying. https://github.com/ethereum/py-evm/pull/964/files/9f8056e83067f6b38a251b8e1cf777043cdd6b87#diff-580a9356192014f1ac5f18afe2008b5aR50

Reply:
Ok, do what you think is best but.. maybe it would be good if the `tx_validation_fn` had a different signature that provided this contextual information?

- `Callable[[block_number, transaction], bool]`
- `Callable[[chain, transaction], bool]`

Reply:
>maybe it would be good if the tx_validation_fn had a different signature that provided this contextual information

I'm not too crazy about that tbh because that seems too much of an implementation detail of a specific validator to me. In general, I'd like to keep the validator fn `Callable[[tx], bool]` because that really seems like the most common interface that all possible validators would share. E.g. imagine you want a validator that takes a proof of authority approach. It wouldn't care about the block number and maybe not even about the chain but instead check if the tx sender is part of some predefined list.

Of course, we could say such a validator may just ignore `chain` or `block_number` as superfluous contextual information but it feels like a smell to me that we don't want it in the interface at all.